core/base/data-updaters/assets-data-updater.php

```diff
@@ -0,0 +1,128 @@
+<?php
+namespace Elementor\Core\Base\Data_Updaters;
+
+use Elementor\Conditions;
+use Elementor\Element_Base;
+use Elementor\Plugin;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+class Assets_Data_Updater extends Document_Data_Updater {
+	const ASSETS_META_KEY = '_elementor_page_assets';
+
+	/**
+	 * @var array
+	 */
+	private $page_assets = [];
+
+	public function update_element( Element_Base $element_data ) {
+		$element_assets = $this->get_element_assets( $element_data );
+
+		if ( $element_assets ) {
+			$this->update_page_assets( $element_assets );
+		}
+	}
+
+	public function is_update_needed() {
+		if ( Plugin::$instance->preview->is_preview_mode() ) {
+			return false;
+		}
+
+		$page_assets = $this->get_page_assets();
+
+		// When $page_assets is array it means that the assets registration has already been made at least once.
+		if ( is_array( $page_assets ) ) {
+			// If $page_assets is not empty then enabling the assets for loading.
+			if ( $page_assets ) {
+				Plugin::$instance->assets_loader->enable_assets( $page_assets );
+			}
+
+			return false;
+		}
+
+		return true;
+	}
+
+	public function after_elements_iteration( $event ) {
+		// Saving the page assets data.
+		$this->document->update_meta( self::ASSETS_META_KEY, $this->page_assets );
+
+		if ( 'render' === $event && $this->page_assets ) {
+			Plugin::$instance->assets_loader->enable_assets( $this->page_assets );
+		}
+	}
+
+	private function get_page_assets() {
+		return $this->document->get_meta( self::ASSETS_META_KEY );
+	}
+
+	private function update_page_assets( $new_assets ) {
+		foreach ( $new_assets as $assets_type => $assets_type_data ) {
+			if ( ! isset( $this->page_assets[ $assets_type ] ) ) {
+				$this->page_assets[ $assets_type ] = [];
+			}
+
+			foreach ( $new_assets[ $assets_type ] as $asset_name ) {
+				if ( ! isset( $this->page_assets[ $assets_type ][ $asset_name ] ) ) {
+					$this->page_assets[ $assets_type ][] = $asset_name;
+				}
+			}
+		}
+	}
+
+	private function get_element_assets( Element_Base $element ) {
+		$controls = $element->get_controls();
+		$settings = $element->get_active_settings();
+		$element_assets = [];
+
+		foreach ( $settings as $setting_key => $setting ) {
+			if ( ! isset( $controls[ $setting_key ] ) ) {
```

Comment:
It's impossible to have an active setting without a corresponding control

Reply:
Done (removed).